sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java

```diff
@@ -0,0 +1,214 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.VarIntCoder;
+import org.apache.beam.sdk.coders.VarLongCoder;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.util.TimeDomain;
+import org.apache.beam.sdk.util.Timer;
+import org.apache.beam.sdk.util.TimerSpec;
+import org.apache.beam.sdk.util.TimerSpecs;
+import org.apache.beam.sdk.util.state.BagState;
+import org.apache.beam.sdk.util.state.StateSpec;
+import org.apache.beam.sdk.util.state.StateSpecs;
+import org.apache.beam.sdk.util.state.ValueState;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link PTransform} that allows to compute elements in batch of desired size. The input {@link
+ * PCollection} needs to be a {@code PCollection<KV>}. Elements that must belong to the same batch
+ * need to have the same key. Elements are added to a buffer. When the buffer reaches {@code
+ * batchSize}, it is then processed through a user {@link SimpleFunction perBatchFn} function. The
+ * output elements then are added to the output {@link PCollection}. Windows are preserved (batches
+ * contain elements from the same window). Batching is done trans-bundles (batches may contain
+ * elements from more than one bundle)
+ *
+ * <p>Example (batch call a webservice and get return codes)
+ *
+ * <pre>{@code
+ * SimpleFunction<Iterable<String>, Iterable<String>> perBatchFn =
+ * new SimpleFunction<Iterable<String>, Iterable<String>>() {
+ *   {@literal @}Override
+ *   public Iterable<String> apply(Iterable<String> input) {
+ *     ArrayList<String> results = callWebService(input);
+ *     return results;
+ *     }
+ *  };
+ *  ...
+ *  Pipeline pipeline = Pipeline.create(...);
+ *  ...
+ *  long batchSize = 100L;
+ *  pipeline.apply(BatchingParDo.via(batchSize, perBatchFn))
+ *          .setCoder(StringUtf8Coder.of());
+ *  pipeline.run();
+ * }</pre>
+ * *
+ */
+public class BatchingParDo<K, InputT, OutputT>
+    extends PTransform<PCollection<KV<K, InputT>>, PCollection<KV<K, OutputT>>> {
+
+  private final long batchSize;
```

Comment:
I'm forgetting here that it's not possible to flush at the end of the bundle and maintain proper windowing. You're right, ignore me.